pipeline/matching/generators/clustering.py

```diff
@@ -0,0 +1,86 @@
+import math
+from typing import Iterator
+
+from pipeline.matching.core import MatchGenerator
+from pipeline.types import Match, MatchingInput, MatchMetadata
+
+GENERATOR_CLUSTERING = "clusteringGenerator"
+
+# TODO: Add type annotations to auxilary functions
+def cosine_similarity(v1, v2):
+    "compute cosine similarity of v1 to v2: (v1 dot v2)/{||v1||*||v2||)"
+    sumxx, sumxy, sumyy = 0, 0, 0
+    for i in range(len(v1)):
+        x = v1[i]
+        y = v2[i]
+        sumxx += x * x
+        sumyy += y * y
+        sumxy += x * y
+    if sumxx * sumyy == 0:
+        return 0
+    return sumxy / math.sqrt(sumxx * sumyy)
+
+
+def closest_match(user1, users):
+    """
+    Summary:
+        loop over all other users to find match with closest aka max cosine similarity --
+        predefined function "cosine_similarity(user1[1],user2[1])"
+        think of users as an 11 dimensional point and trying to find the closest distance between two points
+    Performance: O(?)
+    """
+    cos_sim = 0
+    user1_match = ""
+    for user in users:
+        if user[0] == user1[0]:
+            continue
+        if cosine_similarity(user[1], user1[1]) > cos_sim:
+            cos_sim = cosine_similarity(user[1], user1[1])
+            user1_match = user[0]
+    return user1_match
+
+
+def converts_user_data_to_vector(users_w_data):
+    """
+    Summary:
+    Performance:
+    """
+    intents = ["involve-iit", "watch-recs", "tech-careers"]
+    interests = ["art", "chess", "fashion", "movies", "sports"]
+    vectors = []
+    user_vector = [0] * 11
+    for user in users_w_data:
+        for interest in user.interests:
+            for i in range(5):
+                if interest.name == interests[i]:
+                    user_vector[i] = 1
+        for intent in user.intents:
+            c = 0
+            for i in range(5, 11):
+                if intent.name == intents[c]:
+                    if (i % 2 == 1) and intent.side.SEEKING:
+                        user_vector[i] = 1
+                    if (i % 2 == 0) and intent.side.GIVING:
+                        user_vector[i] = 1
+                if i % 2 == 0:
+                    c += 1
```

Comment:
I know you did not want that many if statements so you wrote a for loop, this gave you some more python practice.  However, this is a little harder to read if you don't know what is going to add so make sure to explain this in your documentation.